personalities/sctfe/storage.go

```diff
@@ -94,33 +95,38 @@ func (cts *CTStorage) AddIssuerChain(ctx context.Context, chain []*x509.Certific
 // Only up to N keys will be stored locally.
 // TODO(phboneff): add monitoring for the number of keys
 type cachedIssuerStorage struct {
-	m map[string]bool
+	sync.RWMutex
+	m map[string]struct{}
 	N int // maximum number of entries allowed in m
 	s IssuerStorage
 }
 
 // Exists checks if the key exists in the local cache, if not checks in the underlying storage.
 // If it finds it there, caches the key locally.
-func (c cachedIssuerStorage) Exists(ctx context.Context, key []byte) (bool, error) {
+func (c *cachedIssuerStorage) Exists(ctx context.Context, key []byte) (bool, error) {
```

Comment:
Is `Exists` meant to be used by higher level code?

The implementation of `Exists` and `AddIssuers` are not really all that different; I wonder if you could simply higher level code just call `AddIssuers` (`AddIssuersIfNotExist`?) for every submission, and modify that func to also check for the entry in `m` at the top of the loop, and only call the underlying storage Add() iff there is anything to store.

As it is, if a new issuer is seen it looks like you'll end up taking the lock and (re-)writing the same entry to `m` 2 or possibly 3 times:
- once in a higher level direct call to `Exists` (L119-121 if the issuer is new and is not in `m` or in the storage) (assuming that's supposed to happen)
- once via a call to `Exists` from `AddIssuers` once the higher level code sees that the issuer doesn't exist and wants to add it
- once inside `AddIssuers` at L119-121

Reply:
Oke doke - can you pop a TODO in there?

Reply:
Sure thing, PR here: https://github.com/phbnf/trillian-tessera/compare/localcache...phbnf:trillian-tessera:inline?expand=1